File: src/brains/recommend/movielens.rb
#!/usr/bin/env ruby

class MovieLens

  attr_reader :movie_ratings, :user_ratings, :movies

  def initialize(users, movies, ratings)
    @user_ratings = {}
    @movie_ratings = {}
    @movies = {}

    open(users).each do |u|
      id, *rest = u.split("::")
      @user_ratings[id.to_i] = {}
    end
    open(movies).each do |m|
      id, title, *rest = m.split("::")
      @movie_ratings[id.to_i] = {}
      @movies[id.to_i] = title
    end
    open(ratings).each do |m|
      userid, movieid, rating = m.split("::").collect { |i| i.to_i }
      @user_ratings[userid][movieid] = rating
      @movie_ratings[movieid][userid] = rating
    end
  end

  def recommend(user)
  end

  # calculate similarity of two movies via pearson correlation
  def similarity(a, b)
    a_ratings = @movie_ratings[a] || []
    b_ratings = @movie_ratings[b] || []
    return 0.0 if a_ratings.empty? or b_ratings.empty?
    mean_all_a = mean(a_ratings.values)
    mean_all_b = mean(b_ratings.values)
    common_a = []
    common_b = []

    # find ratings of other movies by same user
    # subtract mean and store deltas
    @movie_ratings[a].each do |user,rating_a|
      rating_b = @movie_ratings[b][user]
      if rating_b
        common_a << rating_a - mean_all_a
        common_b << rating_b - mean_all_b
      end
    end
    return 0.0 if common_a.size < 3
    pearson_correlation(common_a, common_b)
  end

  def estimate_rating(user, movie)
    rating = movie_ratings[user][movie]
    return rating if rating

    weighted_sum = 0.0
    num = 0
    user_ratings[user].each do |m,rating|
      sim = similarity(movie, m)
      weight = sim * rating
      weighted_sum += weight
      num += 1
    end
    
    if num > 0
      weighted_sum / num
    else
      0.0
    end
  end

  def pearson_correlation(a, b)
    return 0.0 if a.size == 0 # no overlapping ratings

    xy = 0.0
    mean_a, dev_a = mean_and_standard_deviation(a)
    mean_b, dev_b = mean_and_standard_deviation(b)

    a.size.times do |i|
      xy += (a[i] - mean_a) * (b[i] - mean_b)
    end

    if dev_a.zero? || dev_b.zero?
      ind_a = ind_b = 0.0
      (1 .. a.size - 1).each do |i|
        ind_a += a[i - 1] - a[i]
        ind_b += b[i - 1] - b[i]
      end

      if ind_a.zero? && ind_b.zero?
        # degenerate correlation, all points the same
        return 1.0
      elsif dev_a.zero?
        # otherwise either a or b vary
        dev_a = dev_b
      else
        dev_b = dev_a
      end
    end
      
    xy / (a.size * dev_a * dev_b)
  end

  def mean(values)
    values.inject {|sum,x| sum + x } / values.size
  end

  def mean_and_standard_deviation(values)
    m = mean(values)
    sigma = 0.0
    values.each do |v|
      sigma += (v - m) ** 2
    end
    sigma = sigma / values.size
    dev = Math.sqrt(sigma)
    [m, dev]
  end

end

$stderr.puts "loading"
set = MovieLens.new(*ARGV)
user = 1
ratings = []
$stderr.puts "calculating sims"
set.movie_ratings.keys[0..10].each do |m|
  rating = set.estimate_rating(user, m)
  ratings << [rating, set.movies[m]] unless rating.zero?
end
ratings.sort!
puts "estimated ratings for user #{user}"
ratings.each { |i| printf("%.04f %s\n", *i) }

File: scala/src/main/scala/brains/recommend/MovieLens.scala
package brains.recommend

import collection.mutable.{ HashMap => MMap }
import annotation.tailrec
import io.BufferedSource

class MovieLens(movieFile:BufferedSource, ratingFile:BufferedSource) {

  type Ratings = Map[Int,Map[Int,Double]]

  val SEPARATOR = "::"

  val movies = movieFile.getLines.map { line =>
    val id :: title :: rest = line.split(SEPARATOR).toList
    id.toInt -> title
  }.toMap

  val movieRatings = loadRatings(ratingFile.getLines, Map.empty[Int,Map[Int,Double]])

  @tailrec
  private def loadRatings(file:Iterator[String], ratings:Ratings):Ratings = {
    if (file.isEmpty) ratings
    else {
      val user :: movie :: rating :: rest = file.next.split(SEPARATOR).map(_.toInt).toList
      val oldRatings = ratings.getOrElse(movie, Map.empty[Int,Double])
      val newRating = oldRatings.updated(user, rating.toDouble)
      loadRatings(file, ratings.updated(movie, newRating))
    }
  }

  def similarity(a:Int, b:Int):Double = {
    val ratingsA = movieRatings.get(a).getOrElse { return 0.0 }
    val ratingsB = movieRatings.get(b).getOrElse { return 0.0 }

    val meanAllA = mean(ratingsA.values.toArray)
    val meanAllB = mean(ratingsB.values.toArray)

    // find ratings of both movies by the same users, subtract mean and store deltas
    val commonUsers = ratingsA.keySet & ratingsB.keySet
    if (commonUsers.size < 3) return 0.0
    val commonRatings = commonUsers.toSeq.map(u => (ratingsA(u) - meanAllA, ratingsB(u) - meanAllB))
    val (commonA, commonB) = commonRatings.unzip

    pearson(commonA, commonB)
  }

  def pearson(a:Seq[Double], b:Seq[Double]):Double = {
    if (a.isEmpty) return 0.0 // no overlapping ratings

    var (meanA, devA) = meanAndDev(a)
    var (meanB, devB) = meanAndDev(b)
    val xy = (a zip b).map{ case (a,b) => (a - meanA) * (b - meanB) }.sum

    (devA, devB) match {
      case (0.0, 0.0) => return 1.0 // degenerate correlation, all points the same
      case (0.0, devB) => xy / (a.size * devB * devB)
      case (devA, 0.0) => xy / (a.size * devA * devA)
      case (devA, devB) => xy / (a.size * devA * devB)
    }
  }

  def mean(values:Seq[Double]):Double = values.sum / values.size

  def meanAndDev(values:Seq[Double]):(Double,Double) = {
    val meanV = mean(values)
    val squares = values.map(v => (v - meanV) * (v - meanV)).reduceLeft(_ + _)
    meanV -> math.sqrt(squares / values.size)
  }
}

object MovieLens {
  def openFile(path:String) = new io.BufferedSource(new java.io.FileInputStream(path))

  def main(args:Array[String]) = {
    val set = new MovieLens(openFile(args(0)), openFile(args(1)))
    val movies = set.movies.keySet.toSeq
    val sims = for (m <- movies) yield (set.similarity(86,m), set.movies(m))
    for ((sim, title) <- sims.sorted) {          
      printf("%3f %s\n", sim, title)
    }
  }
}
